core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +99,39 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getTrait(RelCollationTraitDef.INSTANCE);
+    if (collation == null || collation.getFieldCollations().size() == 0) {
+      return null;
+    }
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+
+    // Determine mapping between project input and output fields. If sort
+    // relies on non-trivial expressions, we can't push.
+    for (RelFieldCollation fc : collation.getFieldCollations()) {
+      if (map.getTargetOpt(fc.getFieldIndex()) < 0) {
+        return null;
+      }
+      final RexNode node = getProjects().get(fc.getFieldIndex());
+      if (node.isA(SqlKind.CAST)) {
+        // Check whether it is a monotonic preserving cast, otherwise we cannot push
+        final RexCall cast = (RexCall) node;
+        RelFieldCollation newFc = Objects.requireNonNull(RexUtil.apply(map, fc));
+        final RexCallBinding binding =
+            RexCallBinding.create(getCluster().getTypeFactory(), cast,
+                ImmutableList.of(RelCollations.of(newFc)));
+        if (cast.getOperator().getMonotonicity(binding) == SqlMonotonicity.NOT_MONOTONIC) {
+          return null;
+        }
+      }
+    }
+
+    final RelCollation newCollation =
+        getCluster().traitSet().canonize(RexUtil.apply(map, collation));
```

Comment:
Use `collation.apply(map)`. 
No need to canonize, it is already canonized when you get it.